core-common/src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -420,42 +423,86 @@ private void logFailedEvents(final String topicId, final String eventType, final
+    /**
+     * The method sends list of events to Kafka and waiting for the result of each event.
+     *
+     * @param nakadiRecords list of the events to publish
+     * @return empty list if no errors otherwise list with the errored events
+     */
+    public List<NakadiRecordMetadata> sendEvents(final String topic,
+                                                 final List<NakadiRecord> nakadiRecords) {
         final Producer<byte[], byte[]> producer = kafkaFactory.takeProducer();
+        final CountDownLatch latch = new CountDownLatch(nakadiRecords.size());
+        final Map<NakadiRecord, NakadiRecordMetadata> responses = new ConcurrentHashMap<>();
         try {
-            final ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(
-                    nakadiRecord.getTopic(),
-                    nakadiRecord.getPartition(),
-                    nakadiRecord.getEventKey(),
-                    nakadiRecord.getData());
-
-            // fixme avro do not forget EventOwnerHeader, set via metadata?
-            // for the internal KPI evnets it is fine not to have it
-
-            producerRecord.headers().add(
-                    NakadiRecord.HEADER_FORMAT,
-                    nakadiRecord.getFormat());
-
-            producer.send(producerRecord, ((metadata, exception) -> {
-                if (null != exception) {
-                    LOG.warn("Failed to publish to kafka topic '{}' event-type '{}'",
-                            nakadiRecord.getTopic(), nakadiRecord.getEventType(), exception);
-                    if (isExceptionShouldLeadToReset(exception)) {
-                        kafkaFactory.terminateProducer(producer);
+            for (final NakadiRecord nakadiRecord : nakadiRecords) {
+                final ProducerRecord<byte[], byte[]> producerRecord = new ProducerRecord<>(
+                        topic,
+                        nakadiRecord.getPartition(),
+                        nakadiRecord.getEventKey(),
+                        nakadiRecord.getData());
+
+                // fixme avro do not forget EventOwnerHeader, set via metadata?
+                // for the internal KPI evnets it is fine not to have it
+
+                producerRecord.headers().add(
+                        NakadiRecord.HEADER_FORMAT,
+                        nakadiRecord.getFormat());
+                producer.send(producerRecord, ((metadata, exception) -> {
+                    latch.countDown();
+                    if (exception != null) {
+                        final NakadiRecordMetadata record = new NakadiRecordMetadata(
+                                nakadiRecord.getEventType(),
+                                nakadiRecord.getPartition(),
+                                exception);
+                        responses.put(nakadiRecord, record);
+                    } else {
+                        responses.put(nakadiRecord, NakadiRecordMetadata.NULL_RECORD);
```

Comment:
@adyach Do you think it's worth to shortly document the meaning of the `NULL_RECORD`, so that it's clear in the `prepareResponse()` code?